lib/puppet/transaction/event_manager.rb

```diff
@@ -144,18 +144,33 @@ def process_callback(resource, callback, events)
       return false
     end
 
+    message = "Triggered '#{callback}' from #{events.length} events"
+    event = create_callback_status_event(resource, callback, message) if not resource.is_a?(Puppet::Type.type(:whit))
+
     resource.send(callback)
 
-    if not resource.is_a?(Puppet::Type.type(:whit))
-      resource.notice "Triggered '#{callback}' from #{events.length} events"
-    end
+    resource.notice message if not resource.is_a?(Puppet::Type.type(:whit))
     return true
   rescue => detail
-    resource.err "Failed to call #{callback}: #{detail}"
+    resource_error_message = "Failed to call #{callback}: #{detail}"
+    resource.err resource_error_message
+    if not resource.is_a?(Puppet::Type.type(:whit))
+      event = create_callback_status_event(resource, callback, resource_error_message)
+      event.status = "failure"
+    end
 
     transaction.resource_status(resource).failed_to_restart = true
     resource.log_exception(detail)
     return false
+  ensure
+    if event
+      transaction.resource_status(resource) << event
+    end
+  end
+
+  def create_callback_status_event(resource, callback, message)
+    options = { message: message, status: "success", name: callback }
```

Comment:
Since `callback` is a symbol, we'll want to stringify it explicitly `callback.to_s`. Otherwise in older puppet versions (pre-5.something) we'll serialize it as a symbol unnecessarily.

Reply:
Hmm, looks like we use symbols for event names elsewhere: `:name => :noop_restart`. I bet that was cleaned up in later branches.